doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,299 @@
+
+As seen in ``DetectAppLayerMpmRegister2`` that has ``int progress`` as parameter, and ``DetectAppLayerInspectEngineRegister2``, which expects ``int tx_min_progress``, for instance. In the code snippet,
+``HTTP2StateDataClient``, ``HTTP2StateDataServer`` and ``0`` are the values passed to the functions.
+
+
+.. code-block:: c
+
+    void DetectFiledataRegister(void)
+    {
+        .
+        .
+        DetectAppLayerMpmRegister2("file_data", SIG_FLAG_TOSERVER, 2,
+                PrefilterMpmFiledataRegister, NULL,
+                ALPROTO_HTTP2, HTTP2StateDataClient);
+        DetectAppLayerMpmRegister2("file_data", SIG_FLAG_TOCLIENT, 2,
+                PrefilterMpmFiledataRegister, NULL,
+                ALPROTO_HTTP2, HTTP2StateDataServer);
+        .
+        .
+        DetectAppLayerInspectEngineRegister2("file_data",
+            ALPROTO_HTTP2, SIG_FLAG_TOCLIENT, HTTP2StateDataServer,
+            DetectEngineInspectFiledata, NULL);
+        DetectAppLayerInspectEngineRegister2(
+                "file_data", ALPROTO_FTPDATA, SIG_FLAG_TOSERVER, 0, DetectEngineInspectFiledata, NULL);
+        .
+        .
+    }
+
+_`Progress Tracking`
+====================
+
+As a rule of thumb, transactions will follow a request-response model: if a transaction has had a request and a response, it is complete.
+
+But if a protocol has situations where a request or response won’t expect or generate a message from its counterpart,
+it is also possible to have uni-directional transactions. In such cases, transaction is set to complete at the moment of
+creation.
+
+For example, DNS responses may be considered as completed transactions, because they also contain the request data, so
+all information needed for logging and detection can be found in the response.
+
+In addition, for file transfer protocols, or similar ones where there may be several messages before the file exchange
+is completed (NFS, SMB), it is possible to create a level of abstraction to handle such complexity. This could be achieved by adding phases to the protocol implemented model (e.g., protocol negotiation phase (SMB), request parsed (HTTP), and so on).
+
+This is controlled by implementing states. In Suricata, those will be enums that are incremented as the parsing
+progresses. A state will start at 0. The higher its value, the closer the transaction would be to completion.
+
+The engine interacts with transactions state using a set of callbacks the parser registers. State is defined per flow direction (``STREAM_TOSERVER`` / ``STREAM_TOCLIENT``).
+
+In Summary - Transactions and State
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Initial state value: ``0``.
+- Simpler scenarios: state is simply an int.  ``1`` represents transaction completion, per direction.
+- Complex Transaction State in Suricata: ``enum`` (Rust: ``i32``). Completion is indicated by the highest enum value (some examples are: SSH, HTTP, DNS, SMB).
```

Comment:
HTTP2 also, they have a state diagram in their RFC

Reply:
This makes me realize I'm confused with DNS. That's because I'm mentioning it here as an example of a complex transaction state, while I also mention it before as an instance of a simpler protocol. Is that ok, could others get confused by that, as well? In that case, what can we do to dispel such confusion?

Reply:
You can keep DNS as a simplicity example when it fits